File: builder/workspace.py
import functools
import json
import subprocess

from dataclasses import dataclass
from typing import Mapping, List, Sequence, Set


@dataclass
class WorkspaceInformation:
    location: str
    dependencies: Sequence[str]


@functools.lru_cache(maxsize=1)
def _get_workspaces_information_from_yarn() -> Mapping[str, WorkspaceInformation]:
    workspaces_json = json.loads(
        subprocess.check_output(
            ["yarn", "workspaces", "info", "--silent"], stderr=subprocess.DEVNULL
        ).decode()
    )
    return {
        workspace_name: WorkspaceInformation(
            location=workspace_information["location"],
            dependencies=workspace_information["workspaceDependencies"],
        )
        for workspace_name, workspace_information in workspaces_json.items()
    }


def _get_workspace_information(
    workspaces: Mapping[str, WorkspaceInformation], workspace: str
) -> WorkspaceInformation:
    information = workspaces.get(workspace)
    if information is None:
        raise Exception(f"Workspace {workspace} is not found!")
    return information


def _construct_dependency_chain(
    workspaces: Mapping[str, WorkspaceInformation],
    workspace: str,
    dependency_chain: List[str],
    parent_chain: List[str],
    parent_set: Set[str],
    all_visited: Set[str],
) -> None:
    # Check cyclic dependencies.
    if workspace in all_visited:
        if workspace not in parent_set:
            # We reach the end of the chain because we have visited it before.
            return
        parent_chain.append(workspace)
        first_index = parent_chain.index(workspace)
        cyclic_dependency_chain = " -> ".join(parent_chain[first_index:])
        raise Exception(f"Cyclic dependency detected: {cyclic_dependency_chain}.")

    # Check dependencies.
    workspace_dependencies = _get_workspace_information(
        workspaces=workspaces, workspace=workspace
    ).dependencies

    # Visit dependencies.
    all_visited.add(workspace)
    parent_chain.append(workspace)
    parent_set.add(workspace)
    for dependency in workspace_dependencies:
        _construct_dependency_chain(
            workspaces=workspaces,
            workspace=dependency,
            dependency_chain=dependency_chain,
            parent_chain=parent_chain,
            parent_set=parent_set,
            all_visited=all_visited,
        )
    parent_set.remove(workspace)
    parent_chain.pop()
    dependency_chain.append(workspace)


def get_project_workspaces() -> Sequence[str]:
    workspaces = _get_workspaces_information_from_yarn()
    return [workspace for workspace in workspaces.keys() if "common" not in workspace]


def validate_dependency_chain() -> None:
    """
    Throws if there is a cyclic dependency chain.
    """
    workspaces = _get_workspaces_information_from_yarn()
    for workspace in workspaces.keys():
        _construct_dependency_chain(
            workspaces=workspaces,
            workspace=workspace,
            dependency_chain=[],
            parent_chain=[],
            parent_set=set(),
            all_visited=set(),
        )


def get_dependency_chain(workspace: str) -> Sequence[str]:
    dependency_chain: List[str] = []
    _construct_dependency_chain(
        workspaces=_get_workspaces_information_from_yarn(),
        workspace=workspace,
        dependency_chain=dependency_chain,
        parent_chain=[],
        parent_set=set(),
        all_visited=set(),
    )
    return dependency_chain

File: tools/workspace.js
// @ts-check

const { execSync } = require('child_process');

/**
 * @type {Map<string, readonly string[]>}
 */
const workspaceInformation = (() => {
  /**
   * @type {Map<string, readonly string[]>}
   */
  const map = new Map();
  let output = execSync('yarn workspaces info --silent')
    .toString()
    .trim();
  if (output.startsWith('yarn workspaces')) {
    const lines = output.split('\n');
    output = lines.slice(1, lines.length - 1).join('\n');
  }
  const workspacesJson = JSON.parse(output);
  Object.entries(workspacesJson).forEach(([workspaceName, object]) => {
    map.set(workspaceName, object.workspaceDependencies);
  });
  return map;
})();

/**
 * @type {readonly string[]}
 */
const projectWorkspaces = Array.from(workspaceInformation.keys()).filter(
  workspace => !workspace.startsWith('lib-')
);

/**
 * @param {string} workspace
 * @returns {readonly string[]}
 */
const getWorkspaceDependencies = workspace => {
  const information = workspaceInformation.get(workspace);
  if (information == null) {
    throw new Error(`Workspace ${workspace} is not found!`);
  }
  return information;
};

/**
 *
 * @param {string} workspace
 * @returns {readonly string[]}
 */
const getDependencyChain = workspace => {
  /**
   * @type {string[]}
   */
  const dependencyChain = [];
  /**
   * @type {string[]}
   */
  const parentChain = [];
  /**
   * @type {Set<string>}
   */
  const parentSet = new Set();
  /**
   * @type {Set<string>}
   */
  const allVisited = new Set();

  /**
   * @param {string} node
   * @returns {void}
   */
  const visit = node => {
    // Check cyclic dependencies.
    if (allVisited.has(node)) {
      if (!parentSet.has(node)) {
        // We reach the end of the chain because we have visited it before.
        return;
      }
      parentChain.push(node);
      const firstIndex = parentChain.indexOf(node);
      const cyclicDependencyChain = parentChain.slice(firstIndex, parentChain.length).join(' -> ');
      throw new Error(`Cyclic dependency detected: ${cyclicDependencyChain}`);
    }

    // Check dependencies.
    const workspaceDependencies = getWorkspaceDependencies(node);

    // Visit dependencies
    allVisited.add(node);
    parentChain.push(node);
    parentSet.add(node);
    workspaceDependencies.forEach(visit);
    parentSet.delete(node);
    parentChain.pop();
    dependencyChain.push(node);
  };

  visit(workspace);
  return dependencyChain;
};

/**
 * @throws if there is a cyclic dependency chain.
 */
const validateDependencyChain = () =>
  Array.from(workspaceInformation.keys()).forEach(workspace => {
    getDependencyChain(workspace);
    // eslint-disable-next-line no-console
    console.log(`No cyclic dependency detected with ${workspace} as root.`);
  });

module.exports = { projectWorkspaces, getDependencyChain, validateDependencyChain };
